compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,211 @@
+            throw IllegalStateException("Type argument mismatch between immutable and mutable class")
+        }
+        return mutableClass.typeParameters.map { it.symbol }.zip(targetTypeList).toMap()
+    }
+
+    private data class StubsForCollectionClass(
+        val readOnlyClass: IrClassSymbol,
+        val mutableClass: IrClassSymbol,
+        val mutableOnlyMethods: Collection<IrSimpleFunction>
+    )
+
+    private val preComputedStubs: Collection<StubsForCollectionClass> by lazy {
+        with(context.ir.symbols) {
+            listOf(
+                collection to mutableCollection,
+                set to mutableSet,
+                list to mutableList,
+                map to mutableMap,
+                mapEntry to mutableMapEntry,
+                iterable to mutableIterable,
+                iterator to mutableIterator,
+                listIterator to mutableListIterator
+            ).map { (readOnlyClass, mutableClass) ->
+                val readOnlyMethodSignatures = readOnlyClass.functions.map { it.owner.toSignature() }.toHashSet()
+                val mutableMethods = mutableClass.functions
+                    .map { it.owner }
+                    .filter { it.toSignature() !in readOnlyMethodSignatures }
+                    .toHashSet()
+                StubsForCollectionClass(readOnlyClass, mutableClass, mutableMethods)
+            }
+        }
+    }
+
+    // Compute stubs that should be generated, compare based on signature
+    private fun findRelevantStubMethods(irClass: IrClass): Set<IrSimpleFunction> {
+        val ourStubsForCollectionClasses = preComputedStubs.filter { (readOnlyClass, mutableClass) ->
+            irClass.superTypes.any { superTypeSymbol ->
+                val superType = superTypeSymbol.classOrNull?.owner
+                if (superType == null) {
+                    false
+                } else {
+                    // We need to generate stub methods for following 2 cases:
+                    // current class's direct super type is a java class or kotlin interface, and is an subtype of an immutable collection
+                    (superType.comesFromJava() || superType.isInterface) && superTypeSymbol.isSubtypeOfClass(readOnlyClass)
+                            && !irClass.symbol.isSubtypeOfClass(mutableClass)
+                }
+            }
+        }
+
+        // do a second filtering to ensure only most relevant classes are included.
+        val redundantClasses = ourStubsForCollectionClasses.filter { (readOnlyClass) ->
+            ourStubsForCollectionClasses.any { readOnlyClass != it.readOnlyClass && it.readOnlyClass.isSubtypeOfClass(readOnlyClass) }
+        }.map { it.readOnlyClass }
+
+        // perform type substitution and type erasure here
+        return ourStubsForCollectionClasses.filter { (readOnlyClass) ->
+            readOnlyClass !in redundantClasses
+        }.flatMap {
```

Comment:
`}.flatMap { (readOnlyClass, mutableClass, mutableOnlyMethods) ->` maybe to get rid of `it` below